Add performance timing properties (requestStart / responseStart / responseEnd)

kuoe0 opened the pull request:
<!-- Please describe your changes on the following line: -->

Add the following properties to performance timing API:
- requestStart
- responseStart
- responseEnd

---

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #10428 (github issue number if applicable).

<!-- Either: -->
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/11792)

<!-- Reviewable:end -->

kuoe0 committed:
Add requestStart property

kuoe0 committed:
Add responseStart / responseEnd property

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @kichjang (or someone else) soon.

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/script/dom/webidls/PerformanceTiming.webidl, components/script/dom/document.rs, components/script/dom/servohtmlparser.rs, components/script/script_thread.rs, components/script/dom/performancetiming.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify script code, but no tests are modified. Please consider adding a test!

nox commented:
@bors-servo try

---

Reviewed 4 of 4 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[components/script/script_thread.rs, line 403 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11792#-KKh54_6dLux2o7vidiQ:-KKh54_7wgoi-u-1Ixq-:634216212) ([raw file](https://github.com/servo/servo/blob/52925f25895a4852cae8851a2767f1ef41dcaa71/components/script/script_thread.rs#L403)):*
> ```Rust
> 
>     /// Navigation Timing properties:
>     /// https://w3c.github.io/navigation-timing/#sec-PerformanceNavigationTiming
> ```

Nit: put the link for `requestStart` specifically.

---

*[components/script/script_thread.rs, line 2043 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11792#-KKh4dVg4KgSAHMhLrJh:-KKh4dVg4KgSAHMhLrJi:-1194055975) ([raw file](https://github.com/servo/servo/blob/52925f25895a4852cae8851a2767f1ef41dcaa71/components/script/script_thread.rs#L2043)):*
> ```Rust
>         }
> 
>         // http://w3c.github.io/navigation-timing/#widl-PerformanceNavigationTiming-requestStart
> ```

This link doesn't work.

http://www.w3.org/TR/resource-timing/#dom-performanceresourcetiming-requeststart

---

*[components/script/script_thread.rs, line 2174 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11792#-KKh4s1gSIJ9UBGf0eXm:-KKh4s1gSIJ9UBGf0eXn:-145221091) ([raw file](https://github.com/servo/servo/blob/52925f25895a4852cae8851a2767f1ef41dcaa71/components/script/script_thread.rs#L2174)):*
> ```Rust
> 
> fn update_with_current_time_ms(marker: &Cell<u64>) {
>     if marker.get() == Default::default() {
> ```

What is this test about?

---

*[components/script/dom/servohtmlparser.rs, line 113 \[r2\]](https://reviewable.io:443/reviews/servo/servo/11792#-KKh5W9BBQr2uCoqUyZA:-KKh5W9CYVZrIwgEI0N5:288210843) ([raw file](https://github.com/servo/servo/blob/e5ebaaa2f8c08009b77c484d2e952adeace168aa/components/script/dom/servohtmlparser.rs#L113)):*
> ```Rust
>         });
> 
>         let response_start: Cell<u64> = Cell::new(Default::default());
> ```

Don't use a `Cell` here, just a `u64`.

---

*[components/script/dom/servohtmlparser.rs, line 179 \[r2\]](https://reviewable.io:443/reviews/servo/servo/11792#-KKh5bEvNMGw_KxbWdkb:-KKh5bEvNMGw_KxbWdkc:288210843) ([raw file](https://github.com/servo/servo/blob/e5ebaaa2f8c08009b77c484d2e952adeace168aa/components/script/dom/servohtmlparser.rs#L179)):*
> ```Rust
>         };
> 
>         let response_end: Cell<u64> = Cell::new(Default::default());
> ```

Don't use a `Cell` here, just a `u64`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11792#-:-KKh5ptz-tQAfDunfiAx:-318615280)*
<!-- Sent from Reviewable.io -->

bors-servo commented:
:hourglass: Trying commit e5ebaaa with merge 40bc37d...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/1719)

nox commented:
@bors-servo retry #11703

bors-servo commented:
:hourglass: Trying commit e5ebaaa with merge 2164af8...

Ms2ger commented:
Let's land https://github.com/servo/servo/pull/11796 first, and please do not add any `TR/` links.

jdm commented:
@bors-servo: try

jdm commented:
@bors-servo: retry

bors-servo commented:
:hourglass: Trying commit e5ebaaa with merge af73388...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1739)

ConnorGBrewster commented:
@highfive stopped posting test failures :(
```
  ▶ Unexpected subtest result in /navigation-timing/idlharness.html:
  └ PASS [expected FAIL] PerformanceTiming interface: attribute requestStart

  ▶ Unexpected subtest result in /navigation-timing/idlharness.html:
  └ PASS [expected FAIL] PerformanceTiming interface: attribute responseStart

  ▶ Unexpected subtest result in /navigation-timing/idlharness.html:
  └ PASS [expected FAIL] PerformanceTiming interface: attribute responseEnd

  ▶ Unexpected subtest result in /navigation-timing/idlharness.html:
  └ PASS [expected FAIL] PerformanceTiming interface: window.performance.timing must inherit property "requestStart" with the proper type (11)

  ▶ Unexpected subtest result in /navigation-timing/idlharness.html:
  └ PASS [expected FAIL] PerformanceTiming interface: window.performance.timing must inherit property "responseStart" with the proper type (12)

  ▶ Unexpected subtest result in /navigation-timing/idlharness.html:
  └ PASS [expected FAIL] PerformanceTiming interface: window.performance.timing must inherit property "responseEnd" with the proper type (13)

  ▶ Unexpected subtest result in /navigation-timing/test_timing_attributes_exist.html:
  └ PASS [expected FAIL] window.performance.timing.requestStart is defined.

  ▶ Unexpected subtest result in /navigation-timing/test_timing_attributes_exist.html:
  └ PASS [expected FAIL] window.performance.timing.responseEnd is defined.

  ▶ Unexpected subtest result in /navigation-timing/test_timing_attributes_exist.html:
  └ PASS [expected FAIL] window.performance.timing.responseStart is defined.
```

Ms2ger commented:
That's good to see, but seems to imply we're not passing any test that checks the values of the properties. Is that because our values are wrong, because there are no tests (but we could write some), or because it's impossible to write a test?

nox commented:
There are some tests, but they don't seem to pass.

nox commented:
I'm highly confused as to how these tests ran at all given they are supposed to be skipped by WPT.

nox commented:
Oh never mind, Ms2ger enabled them in `master` so it's not in that particular PR tree.

nox commented:
The tests use `performance.navigation`, which we don't define.

Ms2ger commented:
I would not be happy with this landing without any kind of functional tests.

nox commented:
@KuoE0 Do you want to try writing a test for what you implemented?

KuoE0 commented:
Hi, I'll try to do that! But I got a bad cold recently, I'll do it later. Sorry about that.

nox commented:
@KuoE0 No problem, your health is more important than whatever tests we want. :)

KuoE0 commented:
Hi @nox,

Currently, we don't have `performance.navigation`. Did you mean we won't use WPT to test it, use another test to do that? Or, should we implement `performance.navigation` first?

KuoE0 commented:
I'm trying to use the following way to create a new test:

```
./mach create-wpt tests/wpt/mozilla/tests/mozilla/timing_attributes_order.html
```

Is that a correct way? Is there some document to read to write the test?

---

BTW, I'm trying to run the test "tests/wpt/mozilla/tests/mozilla/calc.html" by the following command:

```
./mach test-wpt --release tests/wpt/mozilla/tests/mozilla/calc.html
```

And the following error always happened:

```
Error running mach:

    ['test-wpt', '--release', 'tests/wpt/mozilla/tests/mozilla/calc.html']

The error occurred in code that was called by the mach command. This is either
a bug in the called code itself or in the way that mach is calling it.

You should consider filing a bug for this issue.

If filing a bug, please include the full output of mach, including this error
message.

The details of the failure are as follows:

CalledProcessError: Command '[u'/usr/local/bin/openssl', 'ca', '-config', '/var/folders/t2/dx362r6506g22yd7wqwgnn8h0000gn/T/tmpJtWsUs/openssl.cfg', '-batch', '-create_serial', '-keyfile', '/var/folders/t2/dx362r6506g22yd7wqwgnn8h0000gn/T/tmpJtWsUs/cakey.pem', '-passin', 'pass:web-platform-tests', '-selfsign', '-extensions', 'v3_ca', '-in', '/var/folders/t2/dx362r6506g22yd7wqwgnn8h0000gn/T/tmpJtWsUs/careq.pem', '-out', '/var/folders/t2/dx362r6506g22yd7wqwgnn8h0000gn/T/tmpJtWsUs/cacert.pem']' returned non-zero exit status 1

  File "/Users/kuoe0/Works/Mozilla/servo/python/servo/testing_commands.py", line 328, in test_wpt
    return self.wptrunner(run_file, **kwargs)
  File "/Users/kuoe0/Works/Mozilla/servo/python/servo/testing_commands.py", line 348, in wptrunner
    return run_globals["run_tests"](**kwargs)
  File "/Users/kuoe0/Works/Mozilla/servo/tests/wpt/run_wpt.py", line 13, in run_tests
    return run.run_tests(paths=paths, **kwargs)
  File "/Users/kuoe0/Works/Mozilla/servo/tests/wpt/run.py", line 36, in run_tests
    success = wptrunner.run_tests(**kwargs)
  File "/Users/kuoe0/Works/Mozilla/servo/tests/wpt/harness/wptrunner/wptrunner.py", line 155, in run_tests
    env_options) as test_environment:
  File "/Users/kuoe0/Works/Mozilla/servo/tests/wpt/harness/wptrunner/environment.py", line 104, in __enter__
    self.config = self.load_config()
  File "/Users/kuoe0/Works/Mozilla/servo/tests/wpt/harness/wptrunner/environment.py", line 152, in load_config
    key_file, certificate = self.ssl_env.host_cert_path(hosts)
  File "/Users/kuoe0/Works/Mozilla/servo/tests/wpt/web-platform-tests/tools/sslutils/openssl.py", line 351, in host_cert_path
    key, cert = self._generate_host_cert(hosts)
  File "/Users/kuoe0/Works/Mozilla/servo/tests/wpt/web-platform-tests/tools/sslutils/openssl.py", line 371, in _generate_host_cert
    self._generate_ca()
  File "/Users/kuoe0/Works/Mozilla/servo/tests/wpt/web-platform-tests/tools/sslutils/openssl.py", line 332, in _generate_ca
    "-out", cert_path)
  File "/Users/kuoe0/Works/Mozilla/servo/tests/wpt/web-platform-tests/tools/sslutils/openssl.py", line 70, in __call__
    output=stdout)
```

My development environment is OS X 10.10.

jdm commented:
Have you tried reading against master? We merged a change yesterday that could have fixed that.

KuoE0 commented:
@jdm I rebase with latest source code, and it works now. Thank you!

KuoE0 commented:
Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[components/script/script_thread.rs, line 403 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11792#-KKh54_6dLux2o7vidiQ:-KLUKxMpcecZumjCH_ru:1758359711) ([raw file](https://github.com/servo/servo/blob/52925f25895a4852cae8851a2767f1ef41dcaa71/components/script/script_thread.rs#L403)):*
<details><summary><i>Previously, nox (Anthony Ramine) wrote…</i></summary>
> Nit: put the link for `requestStart` specifically.
</details>
Fixed.

---

*[components/script/script_thread.rs, line 2043 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11792#-KKh4dVg4KgSAHMhLrJh:-KLUKvHBBVSKRzJt5Kti:1758359711) ([raw file](https://github.com/servo/servo/blob/52925f25895a4852cae8851a2767f1ef41dcaa71/components/script/script_thread.rs#L2043)):*
<details><summary><i>Previously, nox (Anthony Ramine) wrote…</i></summary>
> This link doesn't work.
> 
> http://www.w3.org/TR/resource-timing/#dom-performanceresourcetiming-requeststart
</details>
Fixed.

---

*[components/script/script_thread.rs, line 2174 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11792#-KKh4s1gSIJ9UBGf0eXm:-KLUKDNc68AswnMY_C3G:-2049542729) ([raw file](https://github.com/servo/servo/blob/52925f25895a4852cae8851a2767f1ef41dcaa71/components/script/script_thread.rs#L2174)):*
<details><summary><i>Previously, nox (Anthony Ramine) wrote…</i></summary>
> What is this test about?
</details>
I refer this function from [document.rs]. I think this test is used to prevent the value is set twice. I think it is reasonable that requestStart only happened once.

[document.rs]: https://github.com/servo/servo/blob/054bb381e94b82fad4e502bc0f3229f39195c713/components/script/dom/document.rs#L2807-L2813

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11792)*
<!-- Sent from Reviewable.io -->

KuoE0 commented:
@nox I updated the patch, could you give me some feedback? And the test, I'm still working on it. Thank you!

nox commented:
The code looks ok, just needs some test it seems. :)

-S-awaiting-review +S-needs-tests 

---

Reviewed 5 of 5 files at r3, 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[components/script/script_thread.rs, line 2174 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11792#-KKh4s1gSIJ9UBGf0eXm:-KLfZoFPiOZZWeVXnG7a:-2097058392) ([raw file](https://github.com/servo/servo/blob/52925f25895a4852cae8851a2767f1ef41dcaa71/components/script/script_thread.rs#L2174)):*
<details><summary><i>Previously, KuoE0 (Chih-Hsuan Kuo (Tommy Kuo)) wrote…</i></summary>
> I refer this function from [document.rs]. I think this test is used to prevent the value is set twice. I think it is reasonable that requestStart only happened once.
> 
> [document.rs]: https://github.com/servo/servo/blob/054bb381e94b82fad4e502bc0f3229f39195c713/components/script/dom/document.rs#L2807-L2813
</details>
Then replace this test by a proper assertion with `assert!()`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11792#-:-KLf_HzT09hHhjz_ZHFc:-1571864023)*
<!-- Sent from Reviewable.io -->

KuoE0 commented:
Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[components/script/script_thread.rs, line 2174 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11792#-KKh4s1gSIJ9UBGf0eXm:-KMmZ_GXelNPQSLOkpSu:1133668408) ([raw file](https://github.com/servo/servo/blob/52925f25895a4852cae8851a2767f1ef41dcaa71/components/script/script_thread.rs#L2174)):*
<details><summary><i>Previously, nox (Anthony Ramine) wrote…</i></summary>
> Then replace this test by a proper assertion with `assert!()`.
</details>
I don't think we should use `assert!()` here. After doing some investigation, I found requestStart updating could happen  twice or more if there is any iframe in the page. So we only need to keep the first value and ignore the others.

I also referred to Gecko ([source code]). In Gecko, it also keeps the first value we set, and ignores others.

[source code]: https://dxr.mozilla.org/mozilla-central/source/netwerk/protocol/http/nsHttpTransaction.cpp#2016-2023
 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11792)*
<!-- Sent from Reviewable.io -->

KuoE0 commented:
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[components/script/dom/servohtmlparser.rs, line 113 \[r2\]](https://reviewable.io:443/reviews/servo/servo/11792#-KKh5W9BBQr2uCoqUyZA:-KMmanLFL14exONqoAaX:-499146009) ([raw file](https://github.com/servo/servo/blob/e5ebaaa2f8c08009b77c484d2e952adeace168aa/components/script/dom/servohtmlparser.rs#L113)):*
<details><summary><i>Previously, nox (Anthony Ramine) wrote…</i></summary>
> Don't use a `Cell` here, just a `u64`.
</details>
Done.

---

*[components/script/dom/servohtmlparser.rs, line 179 \[r2\]](https://reviewable.io:443/reviews/servo/servo/11792#-KKh5bEvNMGw_KxbWdkb:-KMmamWUWG0emBbwcSg0:-499146009) ([raw file](https://github.com/servo/servo/blob/e5ebaaa2f8c08009b77c484d2e952adeace168aa/components/script/dom/servohtmlparser.rs#L179)):*
<details><summary><i>Previously, nox (Anthony Ramine) wrote…</i></summary>
> Don't use a `Cell` here, just a `u64`.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11792)*
<!-- Sent from Reviewable.io -->

kuoe0 committed:
Copy and modify timing_attributes_order.html from WPT

KuoE0 commented:
Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[components/script/script_thread.rs, line 2174 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11792#-KKh4s1gSIJ9UBGf0eXm:-KMqsZU4R16bBkp3UUWh:908401632) ([raw file](https://github.com/servo/servo/blob/52925f25895a4852cae8851a2767f1ef41dcaa71/components/script/script_thread.rs#L2174)):*
<details><summary><i>Previously, KuoE0 (Chih-Hsuan Kuo (Tommy Kuo)) wrote…</i></summary>
> I don't think we should use `assert!()` here. After doing some investigation, I found requestStart updating could happen  twice or more if there is any iframe in the page. So we only need to keep the first value and ignore the others.
> 
> I also referred to Gecko ([source code]). In Gecko, it also keeps the first value we set, and ignores others.
> 
> [source code]: https://dxr.mozilla.org/mozilla-central/source/netwerk/protocol/http/nsHttpTransaction.cpp#2016-2023
>  
</details>
Please ignore the comment above. I think I need to more investigation on the behavior of requestStart.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11792)*
<!-- Sent from Reviewable.io -->

KuoE0 commented:
Hi @nox, would you like to explain how script_thread created to me? I found script_thread is created from https://github.com/servo/servo/blob/master/components/script/script_thread.rs#L468-L470. But I don't know who calls ScriptThreadFactory::create() (https://github.com/servo/servo/blob/master/components/script/script_thread.rs#L449). Is script_thread a singleton?

ConnorGBrewster commented:
@KuoE0 https://dxr.mozilla.org/servo/source/components/constellation/pipeline.rs#444

bors-servo commented:
:umbrella: The latest upstream changes (presumably #12468) made this pull request unmergeable. Please resolve the merge conflicts.

nox commented:
Still got to review the test but I have some questions already.

-S-needs-rebase -S-needs-tests +S-awaiting-answer 

---

Reviewed 5 of 5 files at r5, 5 of 5 files at r6, 5 of 5 files at r7, 5 of 5 files at r8, 1 of 1 files at r9, 1 of 1 files at r10, 5 of 5 files at r11.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/servo/servo/11792#-KO5WUgaXbPNb6Dje31N:-KO5WUgaXbPNb6Dje31O:-1714196214) (no related file):*
@Ms2ger Should the tests go into WPT?

---

*[components/script/script_thread.rs, line 2174 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11792#-KKh4s1gSIJ9UBGf0eXm:-KO5X6QOG-bFRo-wZEN_:139198085) ([raw file](https://github.com/servo/servo/blob/52925f25895a4852cae8851a2767f1ef41dcaa71/components/script/script_thread.rs#L2174)):*
<details><summary><i>Previously, KuoE0 (Chih-Hsuan Kuo (Tommy Kuo)) wrote…</i></summary>
> Please ignore the comment above. I think I need to more investigation on the behavior of requestStart.
</details>
What was the conclusion of your investigation?

---

*[components/script/script_thread.rs, line 402 \[r5\]](https://reviewable.io:443/reviews/servo/servo/11792#-KO5Vz5UAXM9g4FYxkHZ:-KO5Vz5UAXM9g4FYxkH_:207999716) ([raw file](https://github.com/servo/servo/blob/1ff92ace3e557bd7390324ba98bf4b49d81f820e/components/script/script_thread.rs#L402)):*
> ```Rust
>     content_process_shutdown_chan: IpcSender<()>,
> 
>     panic_chan: IpcSender<PanicMsg>,
> ```

Why was that removed?

---

*[components/script/script_thread.rs, line 2239 \[r10\]](https://reviewable.io:443/reviews/servo/servo/11792#-KO5W9wOzSI8mXpHKMfD:-KO5W9wOzSI8mXpHKMfE:754231867) ([raw file](https://github.com/servo/servo/blob/38133b8b0f3e1f70cf25aef5eced6f9db5c9c6c5/components/script/script_thread.rs#L2239)):*
> ```Rust
> fn update_with_current_time_ms(marker: &Cell<u64>) {
> 
>     debug!("update_with_current_time_ms called!");
> ```

Not sure we want to add new arbitrary logging routines in the middle of nowhere.

---

*[tests/wpt/mozilla/meta/mozilla/timing_attributes_order.html.ini, line 2 \[r11\]](https://reviewable.io:443/reviews/servo/servo/11792#-KO5WNzvPPYVyhNyKKlM:-KO5WNzvPPYVyhNyKKlN:-1283541741) ([raw file](https://github.com/servo/servo/blob/0a3b07bc1ea2b60be0a072cc8138d1f38f4623a5/tests/wpt/mozilla/meta/mozilla/timing_attributes_order.html.ini#L2)):*
> ```INI
> [timing_attributes_orfer.html]
>   type: testharness
> ```

No need for a .ini file if the test passes.

---

*[tests/wpt/mozilla/tests/mozilla/resources/webperftestharness.js, line 155 \[r11\]](https://reviewable.io:443/reviews/servo/servo/11792#-KO5X1JXGnNDR62CtwhQ:-KO5X1JYaddof_HuGK1X:1237394920) ([raw file](https://github.com/servo/servo/blob/0a3b07bc1ea2b60be0a072cc8138d1f38f4623a5/tests/wpt/mozilla/tests/mozilla/resources/webperftestharness.js#L155)):*
> ```JavaScript
> {
>     wp_test(function() { assert_true(value !== notequals, msg); }, msg, properties);
> }
> ```

What are these? There are already such functions provided by WPT, no?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11792#-:-KO5XLJsTSC0zupfMbtf:1240132915)*
<!-- Sent from Reviewable.io -->

KuoE0 commented:
Sorry... it's an WIP patch. I think I should put it into another branch. I met some problem (can not launch any page) after rebased to master, and I just fixed it. Please ignore the test case, it is not finished yet.

Sorry for the slow progress, I'm busy on some feature on Fennec this month.

KuoE0 commented:
I move the WIP to another branch now. In current version, the request_start timing for each document is different.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13205) made this pull request unmergeable. Please resolve the merge conflicts.

kuoe0 committed:
Enable requestStart / responseStart / responseEnd attributes exist test in WPT

KuoE0 commented:
Hi @nox, sorry for the late update. Could you review this patch again? I copied the entire test from [[1]]. And I modified it with removing the failed test case.

Because we don't implement `performance.navigation`, this line [[2]] try to get the member of it and caused the JavaScript error. I have no idea about how to disable it without removing it. I tried to use `expected: CRASH`, `expected: ERROR`, and `expected: FAIL`, but they won't work. That's why I copied the entire test case and made some modifications.

Thank you!

[1]: https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/navigation-timing/test_timing_attributes_order.html
[2]: https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/navigation-timing/test_timing_attributes_order.html#L43-L45

bors-servo commented:
:umbrella: The latest upstream changes (presumably #12830) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13742) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
@nox this PR has been waiting on your review.

nox commented:
Oops, sorry for this.

nox commented on components/script/dom/servoparser/mod.rs:
```diff
@@ -366,6 +367,10 @@ impl FetchResponseListener for ParserContext {
 
         self.parser = Some(Trusted::new(&*parser));
 
+        let response_start: u64 = get_current_time_ms();
```
Nit: no need for `: u64`.

nox commented on components/script/dom/servoparser/mod.rs:
```diff
@@ -426,6 +431,10 @@ impl FetchResponseListener for ParserContext {
                 None => return,
             };
+
+            let response_end: u64 = get_current_time_ms();
```
Nit: no need for `: u64`.

nox reviewed (changes requested):
The code looks ok, but given you duplicated the test and not all properties are defined, the expectations are probably wrong, will do a try.

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit f3a8fea with merge 9b49f03...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/586)

nox commented:
```
  ▶ Unexpected subtest result in /navigation-timing/idlharness.html:
  └ PASS [expected FAIL] PerformanceTiming interface: window.performance.timing must inherit property "requestStart" with the proper type (11)

  ▶ Unexpected subtest result in /navigation-timing/idlharness.html:
  └ PASS [expected FAIL] PerformanceTiming interface: window.performance.timing must inherit property "responseStart" with the proper type (12)

  ▶ Unexpected subtest result in /navigation-timing/idlharness.html:
  └ PASS [expected FAIL] PerformanceTiming interface: window.performance.timing must inherit property "responseEnd" with the proper type (13)
```

bors-servo commented:
:umbrella: The latest upstream changes (presumably #14201) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
Closing due to lack of activity.
